eth_p2p/rlpx_protocols/shh.nim

```diff
@@ -0,0 +1,393 @@
+  let bytes = self.toShortRlp()
+
+  var ctx: keccak256
+  ctx.init()
+  ctx.update(bytes)
+
+  var bestPow: float64 = 0.0
+
+  let mineEnd = epochTime() + seconds
+
+  var i: uint64
+  while epochTime() < mineEnd or bestPow == 0: # At least one round
+    var tmp = ctx # copy hash calculated so far - we'll reuse that for each iter
+    tmp.update(i.toBE())
+    i.inc
+    # XXX:a random nonce here would not leak number of iters
+    let pow = calcPow(1, 1, tmp.finish())
+    if pow > bestPow: # XXX: could also compare hashes as numbers instead
+      bestPow = pow
+      result = i.uint64
+
+proc calcPowHash(self: Envelope): Hash =
+  ## Calculate the message hash, as done during mining - this can be used to
+  ## verify proof-of-work
+
+  let bytes = self.toShortRlp()
+
+  var ctx: keccak256
+  ctx.init()
+  ctx.update(bytes)
+  ctx.update(self.nonce.toBE())
+  return ctx.finish()
+
+# Messages ---------------------------------------------------------------------
+
+proc cmpPow(a, b: Message): int =
+  ## Biggest pow first, lowest at the end (for easy popping)
+  if a.pow > b.pow: 1
+  elif a.pow == b.pow: 0
+  else: -1
+
+proc initMessage(env: Envelope): Message =
+  result.env = env
+  result.hash = env.calcPowHash()
+  result.size = env.toRlp().len().uint64 # XXX: calc len without creating RLP
+  result.pow = calcPow(result.size, result.env.ttl, result.hash)
+
+# Queues -----------------------------------------------------------------------
+
+proc initQueue(capacity: int): Queue =
+  result.items = newSeqOfCap[Message](capacity)
+  result.capacity = capacity
+
+proc prune(self: var Queue) =
+  ## Remove items that are past their expiry time
+  let now = epochTime().uint64
+  self.items.keepIf(proc(m: Message): bool = m.env.expiry > now)
+
+proc add(self: var Queue, msg: Message) =
+  ## Add a message to the queue.
+  ## If we're at capacity, we will be removing, in order:
+  ## * expired messages
+  ## * lowest proof-of-work message - this may be `msg` itself!
+
+  if self.items.len >= self.capacity:
+    self.prune() # Only prune if needed
+
+    if self.items.len >= self.capacity:
+      # Still no room - go by proof-of-work quantity
+      let last = self.items[^1]
+
+      if last.pow > msg.pow or
+        (last.pow == msg.pow and last.env.expiry > msg.env.expiry):
+        # The new message has less pow or will expire earlier - drop it
+        self.items.del(self.items.len() - 1)
+
+  self.items.insert(msg, self.items.lowerBound(msg, cmpPow))
+
+when false:
+  rlpxProtocol shh, 6:
+    proc status(p: Peer, values: openarray[KeyValuePair]) =
+      discard
+
+    proc status(p: Peer, values: openarray[KeyValuePair]) =
+      discard
+
+    proc messages(p: Peer, values: openarray[KeyValuePair]) =
+      discard
+
+    proc powRequirement(p: Peer, values: openarray[KeyValuePair]) =
+      discard
+
+    proc topicFilter(p: Peer, values: openarray[KeyValuePair]) =
+      discard
+
+if isMainModule:
```

Comment:
This test will not run in CI. Why not move it to `tests/` folder?

Reply:
yeah, this is just wip.. more convenient during development (faster build), will move for sure when ready